Custom metrics.

jszczepkowski opened the pull request:
Initial implementation of custom metrics.

jszczepkowski committed:
Custom metrics.

Initial implementation of custom metrics.

jszczepkowski commented:
CC @dchen1107

jszczepkowski commented:
CC @dchen1107 @piosz @mwielgus @fgrzadkowski @DirectXMan12

jszczepkowski commented:
@vishh
Can you please take a look? This PR is not finished yet, but most crucial parts are ready.

TODO:
* move custom metrics implementation to a separate sub-folder inside kubelet,
* extend unittest,
* add API verification for the annotation with custom metrics key.

vishh commented:
We need a flag to make custom metrics opt-in for now. Until we are confident of resource usage and stability, lets hold off on enabling it by default.

jszczepkowski commented:
CC @davidopp

jszczepkowski commented:
@vishh  PTAL

The PR is complete and ready for review. I've moved the implementation to a subdirectory in kubelet, added flag which turns on custom metrics (default: false) and extended unittests. There is no verification in API server yet: I'll add it in a subsequent PR.

ncdc commented:
cc @kubernetes/rh-cluster-infra

DirectXMan12 commented on pkg/kubelet/custommetrics/support.go:
```diff
@@ -0,0 +1,150 @@
+		if i == len(endpoint.Names)-1 { // Do not print comma after the last name.
+			comma = ""
+		}
+		buffer.WriteString(fmt.Sprintf("    \"%s%s\"\n", name, comma))
```
Shouldn't this be `"    %q%s\n"`?  Right now, it looks like we'll get an output of  '    "metric_name,"' (comma inside quotes)

jszczepkowski replied:
Yes. Thanks for finding this! Done.

DirectXMan12 commented on pkg/kubelet/custommetrics/support.go:
```diff
@@ -0,0 +1,150 @@
+	}
+	endpoint := endpoints[0]
+	if len(endpoint.Names) == 0 {
+		return fmt.Errorf("names empty!")
```
This should probably be a bit more descriptive ;-)

vishh replied:
+1

jszczepkowski replied:
Done.

DirectXMan12 commented:
I left a couple notes based on the my initial pass.  In general, it looks good, but I haven't gotten a chance to test it yet.

dchen1107 commented:
cc/ @erictune

vishh commented on pkg/kubelet/custommetrics/support.go:
```diff
@@ -0,0 +1,150 @@
+/*
```
nit: add a dpc.go file for this package.
rename this file to `impl.go`

jszczepkowski replied:
done

vishh commented on pkg/kubelet/custommetrics/support.go:
```diff
@@ -0,0 +1,150 @@
+const (
+	// Key for the annotation defining custom metrics for pod.
+	// More info: http://releases.k8s.io/HEAD/docs/proposals/custom-metrics.md#api
+	CustomMetricsAnnotationKey = "metrics.alpha.kubernetes.io/custom-endpoints"
```
These label keys should not be inside of kubelet package. I'd expect them to reside somewhere in the api package.

jszczepkowski replied:
I thought that annotations keys for alpha features are not kept in the api package. Please see: https://github.com/kubernetes/kubernetes/blob/master/pkg/client/leaderelection/leaderelection.go#L71

Still, I don't have a strong opinion and can move it.

jszczepkowski replied:
On a second thought, we also have:
https://github.com/kubernetes/kubernetes/blob/master/pkg/api/unversioned/well_known_labels.go
So, maybe putting it there is more appropriate. Or maybe even create master/pkg/api/unversioned/well_known_annotations.go file? WDYT?

vishh replied:
Hmm. I find that defining API level labels in the Kubelet code to be
strange. How are our users going to identify these labels? Are we hoping
that documentation should solve the discoverability concern?
cc @erictune

On Fri, Dec 11, 2015 at 7:30 AM, Jerzy Szczepkowski <
notifications@github.com> wrote:

> In pkg/kubelet/custommetrics/support.go
> <https://github.com/kubernetes/kubernetes/pull/18295#discussion_r47366605>
> :
>
> > +	"bytes"
> > +	"encoding/json"
> > +	"fmt"
> > +	"io/ioutil"
> > +	"os"
> > +	"path"
> > +
> > +	"github.com/golang/glog"
> > +	"k8s.io/kubernetes/pkg/api"
> > +	kubecontainer "k8s.io/kubernetes/pkg/kubelet/container"
> > +)
> > +
> > +const (
> > +	// Key for the annotation defining custom metrics for pod.
> > +	// More info: http://releases.k8s.io/HEAD/docs/proposals/custom-metrics.md#api
> > +	CustomMetricsAnnotationKey = "metrics.alpha.kubernetes.io/custom-endpoints"
>
> I thought that annotations keys for alpha features are not kept in the api
> package. Please see:
> https://github.com/kubernetes/kubernetes/blob/master/pkg/client/leaderelection/leaderelection.go#L71
>
> Still, I don't have a strong opinion and can move it.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/18295/files#r47366605>.
>

vishh commented on pkg/kubelet/custommetrics/support.go:
```diff
@@ -0,0 +1,150 @@
+// CustomMetricsEndpoint is a structure defining endpoint for pod's custom metric.
+// It is parsed from JSON from the value of the annotation with CustomMetricsAnnotationKey.
+// More info: http://releases.k8s.io/HEAD/docs/proposals/custom-metrics.md#api
+type CustomMetricsEndpoint struct {
```
Shouldn't this structure reside somewhere in the api package? Possibly with versioning information?

jszczepkowski replied:
If we put it in api package and version it, there will be no point in keeping it as an annotation. The idea was to start with lean experimental feature, with flexible API. That's why we keep it in annotation before we get some experience and feedback. After some time, we will promote it to alpha/beta API object with compatibility guarantees.

vishh commented on pkg/kubelet/custommetrics/support.go:
```diff
@@ -0,0 +1,150 @@
+	// More info: http://releases.k8s.io/HEAD/docs/proposals/custom-metrics.md#api
+	CustomMetricsAnnotationKey = "metrics.alpha.kubernetes.io/custom-endpoints"
+
+	CustomMetricsDefinitionContainerPath = "/etc/cadvisor-custom-metrics-prometheus.json"
```
Cadvisor is embedded into the kubelet. I find it silly that we use this file to communicate in-memory. Can you file an issue against cadvisor to expose an APi for configuring custom metrics for containers?

vishh replied:
Maybe filing an issue in kube side to change this behavior will also help us cut down on this technical complexity. 
Someone from the community can also help us reduce this technical debt.

jszczepkowski replied:
I agree. Filled issue: https://github.com/google/cadvisor/issues/1016.

yujuhong replied:
+1 for that. It's weird that we have to jump through the hoops to communicate with cadvisor, and even write this as a docker label (which is not portable). That's the first impression I got when reading this PR...

vishh commented on pkg/kubelet/custommetrics/support.go:
```diff
@@ -0,0 +1,150 @@
+		return nil, nil
+	}
+
+	if err := os.MkdirAll(podContainerDir, 0750); err != nil {
```
Where does this file get cleaned up?

vishh replied:
I'm not happy with adding this mount logic to kubelet. Before we enable this feature by default, let's please enhance the cadvisor API and avoid this mount.

jszczepkowski replied:
pod container dir is a subdir of pod dir, which I guess is cleaned up here: https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/kubelet.go#L1820

vishh commented on pkg/kubelet/custommetrics/support.go:
```diff
@@ -0,0 +1,150 @@
+	}
+}
+
+func MakeCustomMetricsMountIfRequired(
```
I'd prefer an interface for custommetrics package.

jszczepkowski replied:
Currently, I have two public functions (AddCustomMetricsLabelIfRequired and MakeCustomMetricsMountIfRequired) which are not instance methods. Introducing an interface will require making them instance methods that need to be run on a dummy custom metrics package object (object w/o any instance variables). So, in every call, we will need to create an empty custom metric package object just to call a method on it. I don't think this will be cleaner, but if you insist I will do it.

vishh replied:
So cAdvisor is not enabled on all builds. Thats the reason why its
underneath an interface. As long as this code will function irrespective of
whether cAdvisor's existence, the current implementation should be fine.

On Fri, Dec 11, 2015 at 6:11 AM, Jerzy Szczepkowski <
notifications@github.com> wrote:

> In pkg/kubelet/custommetrics/support.go
> <https://github.com/kubernetes/kubernetes/pull/18295#discussion_r47357888>
> :
>
> > +// It is parsed from JSON from the value of the annotation with CustomMetricsAnnotationKey.
> > +// More info: http://releases.k8s.io/HEAD/docs/proposals/custom-metrics.md#api
> > +type CustomMetricsEndpoint struct {
> > +	API   string   `json:"api",omitempty`
> > +	Path  string   `json:"path",omitempty`
> > +	Port  int      `json:"port"`
> > +	Names []string `json:"names"`
> > +}
> > +
> > +func AddCustomMetricsLabelIfRequired(pod *api.Pod, labels map[string]string) {
> > +	if _, ok := pod.Annotations[CustomMetricsAnnotationKey]; ok {
> > +		labels["io.cadvisor.metric.prometheus"] = CustomMetricsDefinitionContainerPath
> > +	}
> > +}
> > +
> > +func MakeCustomMetricsMountIfRequired(
>
> Currently, I have two public functions (AddCustomMetricsLabelIfRequired
> and MakeCustomMetricsMountIfRequired) which are not instance methods.
> Introducing an interface will require making them instance methods that
> need to be run on a dummy custom metrics package object (object w/o any
> instance variables). So, in every call, we will need to create an empty
> custom metric package object just to call a method on it. I don't think
> this will be cleaner, but if you insist I will do it.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/18295/files#r47357888>.
>

vishh commented on pkg/kubelet/custommetrics/support.go:
```diff
@@ -0,0 +1,150 @@
+}
+
+func ensureCustomMetricsFile(fileName, hostIP string, endpoints []CustomMetricsEndpoint) error {
+	// TODO(jsz): This is a temporal cAdvisro limitation, fix it.
```
nit: typo.

vishh replied:
Is there an issue filed for this?

jszczepkowski replied:
We decided that it is fine to have a single endpoint per container. Removed.

vishh commented on pkg/kubelet/kubelet.go:
```diff
@@ -871,17 +880,17 @@ func (kl *Kubelet) initializeModules() error {
 	// Step 4: Start the image manager.
 	if err := kl.imageManager.Start(); err != nil {
-		return fmt.Errorf("Failed to start ImageManager, images may not be garbage collected: %v", err)
+		return fmt.Errorf("failed to start ImageManager, images may not be garbage collected: %v", err)
```
Thanks for the cleanup :+1:

vishh commented on pkg/kubelet/dockertools/manager.go:
```diff
@@ -651,7 +656,7 @@ func (dm *DockerManager) runContainer(
 	// termination information like the termination grace period and the pre stop hooks.
 	// TODO: keep these labels up to date if the pod changes
-	labels := newLabels(container, pod, restartCount)
+	labels := newLabels(container, pod, restartCount, dm.enableCustomMetrics)
```
Why are you storing the custom metrics bit here?

cc @yujuhong

jszczepkowski replied:
We need to add a proper label for cAdvisor for custom metrics. When adding new labels, we need to decide, if custom metrics support is on and if we want to add a special label for custom metrics or not.

vishh commented:
@jszczepkowski: I did my first pass. Apologies for the delay.

jszczepkowski commented:
@vishh Thanks for the review! I've addressed all the comments, PTAL.

DirectXMan12 commented on pkg/kubelet/custommetrics/impl.go:
```diff
@@ -0,0 +1,149 @@
+		for _, endpoint := range endpoints {
+			if port.ContainerPort == endpoint.Port {
+				matching = append(matching, endpoint)
+			}
```
Should we maybe warn if you have a non-matching endpoints?

jszczepkowski replied:
I think It is perfectly fine to have container w/o matching endpoints. Please note that we may have many containers in a pod and only some of them will expose custom metrics and match.

DirectXMan12 replied:
>  Please note that we may have many containers in a pod and only some of them will expose custom metrics and match.

Ah, I see.  I was thinking more at the pod level -- it would be nice to say "you mention custom metrics on a port that you haven't noted on any of this pod's containers -- did you really mean to do that"?

yujuhong commented on pkg/kubelet/custommetrics/impl.go:
```diff
@@ -0,0 +1,149 @@
+	Names []string `json:"names"`
+}
+
+func AddCustomMetricsLabelIfRequired(pod *api.Pod, labels map[string]string) {
```
The package should not have a docker-specific function, and also allows dockertools to call it. Move this function to dockertools/labels.go

jszczepkowski replied:
Done.

yujuhong commented on pkg/kubelet/custommetrics/impl.go:
```diff
@@ -0,0 +1,149 @@
+	numCustomMetrics := 0
+	for _, endpoint := range endpoints {
+		if len(endpoint.Names) == 0 {
+			return nil, fmt.Errorf("custom metrics endpoint definition cannot contain empty names: %s", endpoint)
```
s/%s/+%v

jszczepkowski replied:
done

yujuhong commented on pkg/kubelet/custommetrics/impl.go:
```diff
@@ -0,0 +1,149 @@
+	var endpoints []CustomMetricsEndpoint
+	err := json.Unmarshal([]byte(definition), &endpoints)
+	if err != nil {
+		return nil, fmt.Errorf("error while parsing custom metrics definition '%s': %v", definition, err)
```
s/'%s'/%q

jszczepkowski replied:
done

yujuhong commented on pkg/kubelet/custommetrics/impl.go:
```diff
@@ -0,0 +1,149 @@
+
+func AddCustomMetricsLabelIfRequired(pod *api.Pod, labels map[string]string) {
+	if _, ok := pod.Annotations[CustomMetricsAnnotationKey]; ok {
+		labels["io.cadvisor.metric.prometheus"] = CustomMetricsDefinitionContainerPath
```
@vishh, is label the best way to convey this information to cadvisor?

vishh replied:
Not really. An issue has been filed against cAdvisor for improving this
interface.

On Fri, Dec 11, 2015 at 10:48 AM, Yu-Ju Hong <notifications@github.com>
wrote:

> In pkg/kubelet/custommetrics/impl.go
> <https://github.com/kubernetes/kubernetes/pull/18295#discussion_r47389207>
> :
>
> > +	CustomMetricsDefinitionContainerPath = "/etc/cadvisor-custom-metrics-prometheus.json"
> > +)
> > +
> > +// CustomMetricsEndpoint is a structure defining endpoint for pod's custom metric.
> > +// It is parsed from JSON from the value of the annotation with CustomMetricsAnnotationKey.
> > +// More info: http://releases.k8s.io/HEAD/docs/proposals/custom-metrics.md#api
> > +type CustomMetricsEndpoint struct {
> > +	API   string   `json:"api",omitempty`
> > +	Path  string   `json:"path",omitempty`
> > +	Port  int      `json:"port"`
> > +	Names []string `json:"names"`
> > +}
> > +
> > +func AddCustomMetricsLabelIfRequired(pod *api.Pod, labels map[string]string) {
> > +	if _, ok := pod.Annotations[CustomMetricsAnnotationKey]; ok {
> > +		labels["io.cadvisor.metric.prometheus"] = CustomMetricsDefinitionContainerPath
>
> @vishh <https://github.com/vishh>, is label the best way to convey this
> information to cadvisor?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/18295/files#r47389207>.
>

jszczepkowski replied:
This is the way how curretnly cAvisor works. It was described here: https://github.com/kubernetes/kubernetes/blob/master/docs/proposals/custom-metrics.md#implementation-notes. Custom metrics are alpha feature, so I think we should carry on with this implementation and enhance it in future.

yujuhong replied:
Will we be able to drop the labels in the future once the cadvisor support is in-place, i.e., stop supporting old containers using the labels?

yujuhong commented on pkg/kubelet/kubelet.go:
```diff
@@ -1313,6 +1323,19 @@ func (kl *Kubelet) GenerateRunContainerOptions(pod *api.Pod, container *api.Cont
+	if kl.enableCustomMetrics {
+		customMetricsMount, err := custommetrics.MakeCustomMetricsMountIfRequired(
+			pod, container, kl.getPodContainerDir(pod.UID, container.Name), kl.maxCustomMetricsPerPod)
+		if err != nil {
```
If a user define more custom metrics than the allowed maximum number, the pod will be in the pending phase and SyncPod will retry indefinitely. Is that the intended behavior?

jszczepkowski replied:
Great thanks for spotting this! This is rather not desire behavior. I've added verification of custom metrics to canRunPod function. It should help.

yujuhong commented on pkg/kubelet/custommetrics/impl.go:
```diff
@@ -0,0 +1,149 @@
+	if err := os.MkdirAll(podContainerDir, 0750); err != nil {
+		return nil, fmt.Errorf("error creating pod-container directory: %v", err)
+	}
+	hostsFilePath := path.Join(podContainerDir, "cadvisor-custom-metrics-prometheus.json")
```
`cadvisor-custom-metrics-prometheus.json` needs to be a constant too since it's part of `CustomMetricsDefinitionContainerPath`

jszczepkowski replied:
Done.

yujuhong commented on pkg/kubelet/custommetrics/impl.go:
```diff
@@ -0,0 +1,149 @@
+	buffer.WriteString("  \"metrics_config\" : [\n")
+	for i, name := range endpoint.Names {
+		comma := ","
+		if i == len(endpoint.Names)-1 { // Do not print comma after the last name.
```
nit: use strings.Join() instead?

DirectXMan12 replied:
The issue is that you need to, at the very least, quote everything first.  Additionally, if you want the indentation, that gets a bit weird looking with strings.Join as well.

jszczepkowski replied:
I think strings.Join will not make it more readable.

yujuhong commented on pkg/kubelet/custommetrics/impl.go:
```diff
@@ -0,0 +1,149 @@
+	}
+	endpoint := endpoints[0]
+	if len(endpoint.Names) == 0 {
+		return fmt.Errorf("names of custom metrics to collect cannot be empty")
```
nit: Include the endpoint in the error.

jszczepkowski replied:
done

yujuhong commented on pkg/kubelet/custommetrics/impl.go:
```diff
@@ -0,0 +1,149 @@
+	return matching, nil
+}
+
+func ensureCustomMetricsFile(fileName, hostIP string, endpoints []CustomMetricsEndpoint) error {
```
nit: The function name does not clearly state what it does to me. Perhaps `createCustomMetricsFile` is better?

DirectXMan12 replied:
it *ensures* that the file is present

jszczepkowski replied:
I think ensure is a proper name (*create if needed*). Moreover, it is consistent with convention, see [ensureHostsFile](https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/kubelet.go#L1243) function.

vishh commented:
I thought about this a bit more and chatted with @rjnagal as well. Two separate operations are being combined in this PR.
Ideally, users will include cAdvisor custom metrics configuration inside their images.
HPA being one of the consumers of the custom metrics will then choose the specific metrics that is of interest for the HPA use-case.

Can we require users to embed the custom metrics config into their image? 
How will HPA decide which metrics it needs to query?

jszczepkowski commented:
> Ideally, users will include cAdvisor custom metrics configuration inside their images.
> ... 
> Can we require users to embed the custom metrics config into their image? 

In general, I think there should be two ways of defining custom metrics for pod/container:
A. Embad them in docker image (add file with configuration + label).
B. Use kubernetes API.

Case B was proposed in https://github.com/kubernetes/kubernetes/blob/master/docs/proposals/custom-metrics.md#api. Everyone agreed that we need a easy way to define custom metrics from k8s level. Case A needs playing with docker image, is low level and bit hacky.

Case A is not currently working because kubelet is resetting docker labels.

This PR implement case B. I'm not against case A, we may still leave it as a valid alternative (after fixing resetting labels issue). 

> How will HPA decide which metrics it needs to query?

Here is the proposal https://github.com/kubernetes/kubernetes/issues/18352

jszczepkowski commented:
@vishh @yujuhong 
All comments are addressed, PTAL

dchen1107 commented on pkg/kubelet/custommetrics/impl.go:
```diff
@@ -0,0 +1,156 @@
+		}
+		numCustomMetrics += len(endpoint.Names)
+	}
+	if numCustomMetrics > maxCustomMetricsPerPod {
```
Why you make validation so late? Shouldn't validate if the numCustomMetrics exceeds the max at the validation phase without mounting? If it exceeds, kubelet simply failed to start pod with proper reason?

dchen1107 replied:
Ahh, I misread the code above: VerifyCustomMetricsAnnotation also call getCustomMetricsEndpointsForContainer before makeCustomMetricsMount call the same validation code here.

Then why not move this logic and other validation piece to Verify* method? If validated correctly, we has endpoint for makeCustomMetricsMount?

dchen1107 commented on pkg/kubelet/dockertools/labels.go:
```diff
@@ -67,6 +70,11 @@ func newLabels(container *api.Container, pod *api.Pod, restartCount int) map[str
 
+	// Add annotation for cAdvisor to handle custom metrics if required.
+	if enableCustomMetrics {
+		addCustomMetricsLabelIfRequired(pod, labels)
```
Why introducing a new docker label here? Why not passing such config knob directly to cAdvisor since cAdvisor is linked with Kubelet anyway? In your PR, the location and file name are both constants, no need to pass as label. 

I don't like to use docker label here because you are relying on a runtime internal implementation, not API, at least it is not align with rkt.

DirectXMan12 replied:
I suspect he was just trying to make use of the existing cAdvisor functionality (in which any Docker container can expose metrics as such) without having to change cAdvisor itself.  There is some ongoing discussion about an cAdvisor API for configuring the metrics collection (instead of going through a label and a file), but I'm unsure where that is going at the moment.

dchen1107 commented:
Close this one in favor of #19944
